# Songbird
Wavelet based segmentation and copy number Estimation of DLP+ scDNASeq data

# To Do List
- [ ] QC Plotting Functions
  - [x] Heatmap plotter
  - [x] Cell Plotter (reads, segmentation, cn)
  - [ ] Quality Scorer
  - [x] Reads per CN
  - [ ] UMAP and PCA plotter
- [x] minimize dependencies? (Rphenograph is github install only...)
- [ ] Allow to skip WGD & Subclone ID
- [ ] Build out Vignette and Documentation
- [ ] Parameter checks for all user facing functions
- [x] Convert all multicore to pbmcapply
- [x] Write Sonbird creation functions from matrix
- [ ] Check all combinations of byo matrix, and bam only for use with songbird
- [ ] Allow merging of sbird_sce objects
- [ ] QDNASeq T2T for 50bp SE reads
  - [ ] Get 50mer mappability track
  - [ ] Get GC Content Track
  - [ ] Identify Centromeric regions
  - [ ] Create exclusion list [try [exclude ranges](https://dozmorovlab.github.io/excluderanges/)]

# To Done (almost) list!
- [x] CN Caller
  - [x] Wrap bedpe/bam loading into a function 
  - [x] read in bam file using qDNASeq
  - [x] Calculate GC+Map Correction
  - [x] Calculate per cell ploidy estimation
  - [x] Figure out correction factor (try HG38 scAbsolute)
  - [x] Instantiate scExperiment Class and populate with metadata
  - [x] Make UBH Matrix and add to scExperiment Class
  - [x] Cluster cells using breakpoint matrix + Phenograph
  - [x] Correct Ploidy Estimation & Perform WGD Detection
  - [x] Produce Final Quality Score
  - [x] Make Copy Matrix
  - [x] Minimal Vignette
  - [x] Run on both hg38 and T2T
  - [x] Expand Bin Exclusion list with T2T?
  - [x] Allow to skip any ploidy estimation (Default mean ploidy range 2-8)
  - [x] Handle empty bams and bedpes gracefully
  - [x] Add flexible tagmentation overlap (for scAbsolute's silliness)
  - [x] ~~Convert bed handling to data.table~~ (read_table single core is the same speed & minimizes dependencies)
- [x] ~~Allow recalibration of correction factor for ploidy estimation~~ No longer using any correction factor
- [x] Add Parallel processing
# Minimal Vignette

1. Preprocess bulk bam using the script in the `bash` folder
    - Script uses [Sinto](https://timoast.github.io/sinto/basic_usage.html), [GNU Parallel](https://www.gnu.org/software/parallel/), and has been tested with [bedtools 2.30.0](https://github.com/arq5x/bedtools2/releases/tag/v2.30.0), and [samtools 1.16.1](https://github.com/samtools/samtools/releases/tag/1.16) (but should be good back to samtools 1.12 when they rewrote markdup to match picard's version).
    - script will create the individualBams subdirectory inside the output folder
```bash
bash preprocess_sample.sh -b /mnt/data/HEK_Controls/DLPPLUS_Ladder/hTERT-1_139566A/alignment/all_cells_bulk_control.bam -o ~/htert-1/
```
2. Use the following snippet to segment and call copy on the per-cell bam and per-cell bedpe file generated by the preprocessing script.

```R
library(Songbird)
folder = '~/hTert-1/individualBams'
bedpes <- list.files(folder, pattern = 'bedpe$', full.names = T)
bams <- list.files(folder, pattern = 'bam$', full.names = T)
sbird_sce <- Songbird::process.batch(bams = bams, bedpes = bedpes, genome = 'hg38', n_cpu = 36)
sbird_sce <- ploidy_correction(sbird_sce, min_reads = 50000)
sbird_sce <- copyCall(sbird_sce)
sbird_sce <- identify_subclones(sbird_sce)

plot_heatmap(sbird_sce, assay_name = 'copy', row_split = sbird_sce$subclone)
```
