![Drawing of a Bird on a musical staff with a artistic rendition of the algorithm](https://github.com/GastroEsoLab/Songbird/blob/master/banner.jpg)
Songbird is a WGD aware copy number caller specifically designed for _very low_ coverage single cell whole genome sequencing data. It was initially designed for DLP+ , but is extensible to any WGS method that uses Tn5 prior to PCR amplification to ligate library adaptors to genomic fragments. For more details, please see our preprint on [bioRxiv](https://doi.org/10.64898/2025.12.18.693686).

# Quickstart

## Install
Songbird largely uses packages you can get from Bioconductor & CRAN. However, the per-genome qDNASeq annotations require manual install.
```R
BiocManager::install("QDNAseq.hg19")
remotes::install_github("asntech/QDNAseq.hg38@main")
remotes::install_github("GastroEsoLab/QDNAseq.hs1@main")
remotes::install_github("GastroEsoLab/Songbird@main")
```

## Bedpe Generation
  - Requires [Sinto](https://timoast.github.io/sinto/basic_usage.html), [GNU Parallel](https://www.gnu.org/software/parallel/), [bedtools > 2.30.0](https://github.com/arq5x/bedtools2/releases/tag/v2.30.0), and [samtools > 1.16.1](https://github.com/samtools/samtools/releases/tag/1.16).
  - The Script will create the individualBams subdirectory inside the output folder
```bash
bash preprocess_sample.sh -b /mnt/data/HEK_Controls/DLPPLUS_Ladder/hTERT-1_139566A/alignment/all_cells_bulk_control.bam -o ~/htert-1/
```

## Run Songbird
```R
library(devtools)
setwd('~/Songbird/')
load_all()

folder = '~/hTert-1/individualBams'
bedpes <- list.files(folder, pattern = 'bedpe$', full.names = T)
bams <- list.files(folder, pattern = 'bam$', full.names = T)
sbird_sce <- Songbird::process.batch(bams = bams, bedpes = bedpes, genome = 'hg38', n_cpu = 36)
sbird_sce <- ploidy_correction(sbird_sce, min_reads = 50000)
sbird_sce <- copyCall(sbird_sce)
sbird_sce <- identify_subclones(sbird_sce)
```

## Final Plotting
For your convience & to help with validating the copy number calls, we provide a couple of plotting functions
```R
# Plots a heatmap of the copy number state of the whole sample split by subclone
plot_heatmap(sbird_sce, assay_name = 'copy', row_split = sbird_sce$subclone)

# Plots an individual cell for comparing 
plot_cell(parameters)
```

# Parameters
Most of the impactful parameters are in the initial segmentation and ploidy estimation function. Downstream functions have fewer parameters

- **process.batch**: Performs initial ploidy estimation and segmentation on samples
  - bams: `(Required if not providing bedpe)` List of bam files to perform CN calling
  - genome: `(Required)` Genome version for the bams ('hg19', 'hg38', or 'hs1')
  - bedpes: `(Required if not providing bam)` List of bedpe files generated from the bams. Bedpes and bams must be in the same order. Songbird can be run without bedpe files, but will not be able to use novel ploidy estimation algorithm.
  - bin.size: `(Default: 500000)` Bin size in bp. Possible values depend on the available QDNAseq annotation objects, but are typically `1000, 5000, 10000, 15000, 30000, 100000, 500000, 1000000`
  - min_length: `(Default: 50)` Minimum fragment length for ploidy estimation. 50 seems to perform best for Illumina 150bp short read sequencing.
  - max_length: `(Default: NULL)` Maximum fragment length for ploidy estimation. The default (99th percentile of fragment lengths) performs well for Illumina 150bp short read sequencing, but longer fragment length tools may need to specify a smaller value to ensure that the ploidy estimation is performed on a region with similar genome accessibility
  - tag_overlap: `(Default: 10)` Overlap between adjacent fragments from Tn5 tagmentation. Should only be 10 nt unless using a different transposase, or with trimmed sequencing data
  - n_cpu: `(Default: NULL)` Number of cores to split across. Defaults to all cpus which may cause memory issues with higher coverage samples
  - focal_amps: (`Default: TRUE`) Whether your data may contain focal amplifications. Structural variants may cause many reads to start or end at the same coordinates, biasing the CN caller to higher states. Should only be set to false if you are confident that your data lacks any structural variants (ie diploid or immortalized cell lines).

- **ploidy_correction**: Clusters cells to share information across noisy ploidy estimators & detect any perfect WGD
  - sbird_sce: `(Required)` Songbird object generated by _process.batch_
  - min_reads: `(Default: 100000)` Reads per cell threshold used to collect reliable ploidy estimates for averaging & WGD detection
  - k: `(Default: 45)` Number of neighbors for building the graph prior to cluster detection.

- **copyCall**: Integrates ploidy estimate and segmented data to make a final copy number call for each cell
  - sbird_sce: `(Required)` Songbird object generated by _process.batch_ & ploidy corrected with _ploidy_correction_
  - n_cpu: `(Default: NULL)` Number of cores to split across. Copy Calling is quite efficient and fast, so all cores should be best
 
- **identify_subclones**: Performs final subclone identification. Cluster 0 contains all cells which failed to be clustered properly
  - sbird_sce: `(Required)` Songbird object generated by _process.batch_
  - assay: `(Default 'copy')` Which data to use for clustering. Options are `counts`, `reads`, `segmented`, `copy`
  - k: `(Default: 30)` Number of neighbors for building the graph prior to cluster detection.
  - res: `(Default: 'auto')` Resolution parameter for leiden clustering. `'auto'` will automatically tune to maximize modularity
  - method: `(Default: inv_manhattan)` Which distance function to calculate. Can byo by providing a function object
  - min_size: `(Default: 5)` Minimum cluster size. All cells in clusters smaller than this will be lumped into cluster 0
  - seed: (`Default: 1234)` For reproducibility. Pick your favorite number.
  - min_readCount: `(Default: 100000)` Minimum read count for breakpoint identification for clustering
  - column_name: `(Default: 'subclone')` RowData entry name for the subclone. Change if you want to save older clusterings

- **plot_heatmap**: Convienence function for Plotting
  - sce: `(Required)` Songbird object generated by _process.batch_
  - assay: `(Required)` Which data to plot. Options are `counts`, `reads`, `segmented`, `copy`
  - cell_attribs: `(Default: NULL)` Which columns from RowData should be used to annotate the cells
  - rowClust: `(Default: TRUE)` Whether or not to cluster the rows
  - use_raster: `(Default: TRUE)` Whether or not to rasterize the heatmap. Raster heatmaps will plot faster, but lack detail & may hide focal amps
  - bin_attribs: `(Default: NULL)` Which columns from ColData should be used to annotate the cells
  - return: `(Default: FALSE)` Whether or not to return the heatmap object. Useful for plotting with `cowplot::plot_grid`
  - row_title: `(Default: NULL)` Row title to display
  - plot_title: `(Default: NULL)` Overall Plot Title
  - legend_title: `(Default: NULL)` Heatmap legend label. NULL uses the assay name specified
  - show_cellNames: `(Default: FALSE)` Whether or not to add the cell names to the plot
 
- **plot_cell**: Useful plotting function for hand validating CN calls vs reads
  - sbird_sce: `(Required)` Songbird object generated by _process.batch_
  - cell: `(Required)` Cell Name to grab from the sbird single cell experiment object
  - assay: `(Default: 'copy')` Which assay to pull from
  - chr: `(Default: NULL)` Which chromosome to plot. NULL plots all chromosomes
  - return_plot: `(Default: FALSE)` Whether or not to return the ggplot object. Useful for plotting with `cowplot::plot_grid`

# Tips for use

## Minimum Read Depth Requirements
Songbird is able to accurately estimate the true ploidy of a cell up to hexaploid (CN 6). Depending on the cell's ploidy, the tool requires a minimum of 0.0025x (Diploid) to 0.01x (Pentaploid) coverage per cell to get an accurate copy number estimate. For a 150bp Illumina shortread sequencer, that corresponds to 50000 (Diploid) or 200000 (Pentaploid) reads per cell. 

## CN Calling on non-tumor samples
Cells with many highly amplified, small structural variants (<10kbp in size) can cause the ploidy estimator to overestimate. We filter for this by only using the middle 80th percentile of bins by overlap counting for ploidy estimation. However, if you are analyzing a healthy cell, or a cell line you know is absent of focal amplifications, you can turn off this filter by setting the `focal_amps` parameter in the `process.batch` function to `FALSE`. 

## Use with long read sequencing
In the paper we demonstrate accurate ploidy estimation on data sequenced with PacBio long read sequencing. In order to get a proper ploidy estimate we had to limit the longest fragments considered for ploidy estimation to 8kbp. Since Songbird compares the overlapping read density to upstream read density, an extremely long upstream window can cause the algorithm to consider regions with different genome availability for ploidy estimation. This causes inaccurate estimation.



Banner designed by [Jennifer Jones](https://github.com/jejonesu)
